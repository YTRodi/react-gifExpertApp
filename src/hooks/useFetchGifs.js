import { useEffect, useState } from "react";
import { getFetchGifs } from "../helpers/getFetchGifs";

// Custom HOOK
// Funcionan como si fuera functional components, pueden tener efectos, pueden usar reducer, contextos
export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Me permite poder ejecutar código de manera condicional, es decir,
    // solamente quiero que se ejecute esa condición cuando el componente es RENDERIZADO por PRIMERA vez. ( NO PUEDEN SER async )
    // ( Analogía al patrón de diseño Singleton )
    //     1er param = función a ejecutar
    //     2do param = [] de dependencias

    // Si la catergoria cambia entonces se ejecuta de nuevo el useEffect
    useEffect( () => {

        getFetchGifs( category )
            .then( imgs => {

                // console.log(imgs);
                setState( {
                    data: imgs,
                    loading: false
                });

            });

            console.log(category);
            
    }, [category])

    return state; // { data: [], loading: true }

};