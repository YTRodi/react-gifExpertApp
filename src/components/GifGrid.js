import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';

const GifGrid = ( { category } ) => {

    const { data:images , loading } = useFetchGifs( category );
    // data:images = le agregamos un alias a la data, en esta caso las img.
    
    return (
        <>
            <h3 className="animate__animated animate__fadeInDown">{ category }</h3>

            {/* Operación lógica AND, si es true entonces evalua el primero, sino no evalua nada. */}
            { loading && <p className="animate__animated animate__flash">Loading...</p> }

            <div className="card-grid">
                {
                    images.map( img  => (

                        <GifGridItem 
                            key={ img.id }
                            { ...img } // Estoy enviando cada una de las props de las img como una prop independiente.
                            // img={ img } 
                        />

                    ))
                }
            </div>
        </>
    )
}

export default GifGrid;
