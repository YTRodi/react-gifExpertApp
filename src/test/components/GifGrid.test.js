import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import GifGrid from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

// Finjo cualquier llamada a ese archivo, suponer y controlar la info que me va a traer.
jest.mock("../../hooks/useFetchGifs"); 

describe('Pruebas en GifGrid', () => {
    
    const category = 'Tanjirou';

    test('Debe de mostrar el componente correctamente', () => {
        
        // Lo que hace esto es:
        // Cuando se llame a dicha función dentro del componente, va a regresar el obj.
        // Falseo la data, 'falsa implementación'
        useFetchGifs.mockReturnValue({
            // Estado inicial del objeto
            data: [],
            loading: true
        });
        
        const wrapper = shallow( <GifGrid category={ category } /> );

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de mostrar items cuando se cargan imágenes useFetchGifs (Custom HOOK)', () => {
        
        const gifs = [{
            id: 'YTR',
            url: 'https://localhost/cualquier/cosa.png',
            title: 'Cualquier cosa'
        },
        {
            id: 'Msdiofj',
            url: 'https://localhost/otra/cosa.png',
            title: 'Cualquier otra cosa'
        }
        ];

        useFetchGifs.mockReturnValue({
            // Estado inicial del objeto
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category={ category } /> );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'p' ).exists() ).toBe( false ); // No debería de existir el parrafo
        expect( wrapper.find( 'GifGridItem' ).length ).toBe( gifs.length );

    });

});