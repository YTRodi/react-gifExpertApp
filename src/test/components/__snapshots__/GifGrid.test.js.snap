// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
    Tanjirou
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en GifGrid Debe de mostrar items cuando se cargan im√°genes useFetchGifs (Custom HOOK) 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
    Tanjirou
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="YTR"
      key="YTR"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.png"
    />
    <GifGridItem
      id="Msdiofj"
      key="Msdiofj"
      title="Cualquier otra cosa"
      url="https://localhost/otra/cosa.png"
    />
  </div>
</Fragment>
`;
